### ISSUE TITLE
Slice widening that handles errors in std.

### ISSUE THREAD

**Issue by Hejsil** [zig-lang/zig#983]
I use the slice widening syntax (`([]Other)(some_slice)`) quite a lot in my [randomizer](https://github.com/Hejsil/pokemon-randomizer), and I've come across three different methods I use for handling when this operation can fail:

1. **Let it assert:** This is the default for widening, and you want this when widening a slice of the wrong size is a bug. I tend to want this when I'm working with sizes I know at compile time, or when I've ensured elsewhere that the size is valid (examples: [here](https://github.com/Hejsil/pokemon-randomizer/blob/c43b79c72501fc68e371b461314c318afdfb5a77/src/pokemon/gen3.zig#L191) and [here](https://github.com/Hejsil/pokemon-randomizer/blob/c43b79c72501fc68e371b461314c318afdfb5a77/src/pokemon/gen3.zig#L421)).
2. **Return an error:** This is for when you have a range parameter from the user, and when the range is not valid for widening, and error should be returned. The range is defined by some input, so it's a user error if you can't widen, and not a bug on your end. (example: [here](https://github.com/Hejsil/pokemon-randomizer/blob/c43b79c72501fc68e371b461314c318afdfb5a77/src/nds/fs.zig#L265) (I don't use widen here, but I could have)).
3. **Do a narrow widen:** For when you don't care all items won't fit in the resulting widen, and you just want `([]Other)(slice[0..slice.len - (slice.len % @sizeOf(Other))])` to be the resulting widen. This operation cannot fail. (examples: [here](https://github.com/Hejsil/pokemon-randomizer/blob/c43b79c72501fc68e371b461314c318afdfb5a77/src/nds/fs.zig#L278) and [here](https://github.com/Hejsil/pokemon-randomizer/blob/99d622480dbf615ea870435f7bb9bc6a0a6e6875/src/pokemon/index.zig#L232)).

Now, one could just write the if statements that do checks before widening, and do the narrow widening by hand, but both options make the code quite verbose, where I think a function actually makes it easier to read:
```
const w = try mem.widen(slice, Other);
vs
if (slice.len % @sizeOf(Other) != 0)
    return error.InvalidWiden;
const w = ([]Other)(slice);
```
```
const w = mem.widenNarrow(slice, Other);
vs
const w = ([]Other)(slice[0..slice.len - (slice.len % @sizeOf(Other))]);
```

TLDR: I think we should have `std.mem.widen` and `std.mem.widenNarrow`. I think they have enough use for certain applications that it is worth in `std`, and I think they're more readable that their inline counterparts.

**Comment by BraedonWooding** [comment]
I agree 👍
